src/Algebra/Graph/Undirected.hs

```diff
@@ -0,0 +1,907 @@
+'edge' 1 2 < 'edge' 1 1 + 'edge' 2 2
+'edge' 1 2 < 'edge' 1 3@
+'edge' 1 2 == 'edge' 2 1
+
+Note that the resulting order refines the 'isSubgraphOf' relation and is
+compatible with 'overlay' and 'connect' operations:
+
+@'isSubgraphOf' x y ==> x <= y@
+
+@'empty' <= x
+x     <= x + y
+x + y <= x * y@
+-}
+newtype Graph a = UndirectedGraph { graph :: G.Graph a }
+             deriving (Generic, NFData)
+
+instance (Show a, Ord a) => Show (Graph a) where
+  show = show . toSymmetricRelation
+
+{- See Note [Functions for rewrite rules] in 'Algebra.Graph' -}
+
+instance Functor Graph where
+    fmap = fmapR
+
+fmapR :: (a -> b) -> Graph a -> Graph b
+fmapR f g = bindR g (vertex . f)
+{-# INLINE fmapR #-}
+
+-- | __Note:__ this does not satisfy the usual ring laws; see 'Graph' for more
+-- details.
+instance Num a => Num (Graph a) where
+    fromInteger = vertex . fromInteger
+    (+)         = overlay
+    (*)         = connect
+    signum      = const empty
+    abs         = id
+    negate      = id
+
+instance Ord a => Eq (Graph a) where
+    (==) = eqR
+
+instance Ord a => Ord (Graph a) where
+    compare = ordR
+
+-- TODO: Find a more efficient equality check.
+-- TODO: Implement toUndirectedAdjacencyMap.
+-- Check if two graphs are equal by converting them to their adjacency maps.
+eqR :: Ord a => Graph a -> Graph a -> Bool
+eqR x y = toSymmetricRelation x == toSymmetricRelation y
+{-# NOINLINE [1] eqR #-}
+{-# RULES "eqR/Int" eqR = eqIntR #-}
+
+-- Like 'eqR' but specialised for graphs with vertices of type 'Int'.
+eqIntR :: Graph Int -> Graph Int -> Bool
+eqIntR x y = toSymmetricRelation x == toSymmetricRelation y
+
+-- TODO: Find a more efficient comparison.
+-- TODO: Implement toUndirectedAdjacencyIntMap.
+-- Compare two graphs by converting them to their adjacency maps.
+ordR :: Ord a => Graph a -> Graph a -> Ordering
+ordR x y = compare (toSymmetricRelation x) (toSymmetricRelation y)
+{-# NOINLINE [1] ordR #-}
+{-# RULES "ordR/Int" ordR = ordIntR #-}
+
+-- Like 'ordR' but specialised for graphs with vertices of type 'Int'.
+ordIntR :: Graph Int -> Graph Int -> Ordering
+ordIntR x y = compare (toSymmetricRelation x) (toSymmetricRelation y)
+
+instance Applicative Graph where
+    pure  = vertex
+    (<*>) = apR
+
+apR :: Graph (a -> b) -> Graph a -> Graph b
+apR f x = bindR f (<$> x)
+{-# INLINE apR #-}
+
+instance Monad Graph where
+    return = pure
+    (>>=)  = bindR
+
+bindR :: Graph a -> (a -> Graph b) -> Graph b
+bindR g f = foldg empty f overlay connect g
+{-# INLINE [0] bindR #-}
+
+instance Alternative Graph where
+    empty = empty
+    (<|>) = overlay
+
+instance MonadPlus Graph where
+    mzero = empty
+    mplus = overlay
+
+-- | Construct the /empty graph/.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- 'isEmpty'     empty == True
+-- 'hasVertex' x empty == False
+-- 'vertexCount' empty == 0
+-- 'edgeCount'   empty == 0
+-- 'size'        empty == 1
+-- @
+empty :: Graph a
+empty = UndirectedGraph G.empty
```

Comment:
So, you didn't manage to write `coerce G.empty` here?

Reply:
No, GHC gave me the error I posted... It's something with the quantified variables, I don't know if it needs some extension enabled.. but since I'm having problems with the `Data.List.Extra` maybe it's my setup that's not right. Could you try for yourself and check if `coerce G.empty` works fine?

Reply:
Sorry, apparently things are a bit more complicated than I thought. See https://github.com/snowleopard/alga/pull/220#issuecomment-510891564.

Reply:
Thanks for your help! It's all good now!